test/integration/publishing/save-request.test.js

```diff
@@ -31,827 +32,1714 @@ describe('Save statement and delivery and send to CRM and save failure if so', (
+
+      test('should save failure with deliveryId', async () => {
+        await saveRequest(request, reference, EMAIL, reason)
+
+        const delivery = await db.delivery.findOne()
+        const failure = await db.failure.findOne()
+        expect(failure.deliveryId).toBe(delivery.deliveryId)
+      })
+
+      test('should save failure with reason as REJECTED', async () => {
+        await saveRequest(request, reference, EMAIL, reason)
+
+        const failure = await db.failure.findOne()
+        expect(failure.reason).toBe(REJECTED)
+      })
+
+      test('should save failure with failed as SYSTEM_TIME', async () => {
+        await saveRequest(request, reference, EMAIL, reason)
+
+        const failure = await db.failure.findOne()
+        expect(failure.failed).toStrictEqual(SYSTEM_TIME)
+      })
+
+      test('should send message to CRM', async () => {
+        await saveRequest(request, reference, EMAIL, reason)
+
+        expect(mockMessageSender().sendMessage).toHaveBeenCalled()
+        expect(mockMessageSender().closeConnection).toHaveBeenCalled()
+      })
+
+      test('should send 1 message to CRM', async () => {
+        await saveRequest(request, reference, EMAIL, reason)
+
+        expect(mockMessageSender().sendMessage).toHaveBeenCalledTimes(1)
+        expect(mockMessageSender().closeConnection).toHaveBeenCalledTimes(1)
+      })
+
+      test('should send message to CRM with correct content', async () => {
+        await saveRequest(request, reference, EMAIL, reason)
+
+        expect(mockMessageSender().sendMessage).toHaveBeenCalledWith({
+          ...MESSAGE,
+          body: {
+            email: request.email,
+            errorMessage: INVALID_ERROR,
+            frn: request.frn
+          }
+        })
+      })
+    })
+
+    describe('When failure reason is "Not known failure reason"', () => {
+      beforeEach(async () => {
+        reason = 'Not known failure reason'
+      })
+
+      test('should not save statement', async () => {
+        try { await saveRequest(request, reference, EMAIL, reason) } catch {}
+
+        const statement = await db.statement.findAll()
+        expect(statement.length).toBe(0)
+      })
+
+      test('should not save delivery', async () => {
+        try { await saveRequest(request, reference, EMAIL, reason) } catch {}
+
+        const delivery = await db.delivery.findAll()
+        expect(delivery.length).toBe(0)
+      })
+
+      test('should not save failure', async () => {
+        try { await saveRequest(request, reference, EMAIL, reason) } catch {}
+
+        const failure = await db.failure.findAll()
+        expect(failure.length).toBe(0)
+      })
+
+      test('should not send message to CRM', async () => {
+        try { await saveRequest(request, reference, EMAIL, reason) } catch {}
+
+        expect(mockMessageSender().sendMessage).not.toHaveBeenCalled()
+        expect(mockMessageSender().closeConnection).not.toHaveBeenCalled()
+      })
     })
   })
 })
```

Comment:
I think we need to review our test-files structure, they are becoming to long, hence becoming difficult to review.

Reply:
The fact that our tests are becoming too long is a sign that they are decomposed well enough.

I've aided the ability to review by creating describe blocks where each describe contains only for that document type or that particular scenario. Using VS Code it is easy to 'close these up' to focus on certain ones.

I also believe we're massively under utilising `jest`.

Jest provides a `test.each` functionality (amongst others: https://jestjs.io/docs/api) which allows for dynamic tests to be created for multiple scenarios given a set of inputs. We have somewhat half-heartedly used this in the past: https://github.com/DEFRA/ffc-pay-processing/blob/f069336ecdef040f043bf9d75b3c76f64d806cb9/test/integration/narrow/server/routes/holds.test.js#L145.

I believe that with my describe and sub scenarios decompression this can be accomplished.